File: 01/01.rb
#!/usr/bin/env ruby
first_list = []
second_list = []
File.open("input") do |f|
  f.each_line do |line|
    a, b = line.strip.split("   ")
    first_list << a.to_i
    second_list << b.to_i
  end
end
first_list.sort!
second_list.sort!
# Part one
puts first_list.zip(second_list).map {(_1 - _2).abs}.sum

# Part two
frequencies = second_list.group_by(&:itself).transform_values(&:size)
puts first_list.map {_1 * (frequencies[_1] || 0)}.sum

File: 01/01.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let contents = fs::read_to_string("input").expect("failed to read file 'input'");
    let mut first_table: Vec<u32> = vec![];
    let mut second_table: Vec<u32> = vec![];
    for line in contents.trim_end().split('\n') {
        let (a, b) = line.split_once("   ").expect("unexpected line shape");
        first_table.push(a.parse().expect("unable to read first number"));
        second_table.push(b.parse().expect("unable to read second number"));
    }
    first_table.sort();
    second_table.sort();
    // Part 1
    let sum = first_table.iter().zip(second_table.iter()).map(|(a, b)| a.abs_diff(*b) as u64).reduce(|acc, e| acc + e).unwrap();
    println!("{}", sum);

    // Part 2
    let mut frequencies = HashMap::new();
    for b in second_table {
        frequencies.entry(b).and_modify(|e| *e += 1).or_insert(1);
    }
    let result = first_table.iter().map(|a| a * frequencies.get(a).unwrap_or(&0)).reduce(|acc, e| acc + e).unwrap();
    println!("{}", result);
}
